specs/networking.md

```diff
@@ -19,6 +20,35 @@
 |----------|-----------------------------------------|------------------|
 | `shares` | [Share](./data_structures.md#share)`[]` | Shares in a row. |
 
+### ConsensusProposal
+
+Defined as `ConsensusProposal` [here](./proto/consensus.proto).
+
+When receiving a new block proposal `proposal` from the network, the following steps are performed in order. _Must_ indicates that peers must be blacklisted (to prevent DoS attacks) and _should_ indicates that the network message can simply be ignored.
+
+1. `proposal.type` must be a `SignedMsgType`.
+1. `proposal.round` is processed identically to Tendermint.
+1. `proposal.pol_round` is processed identically to Tendermint.
+1. `proposal.header` must be well-formed.
+1. `proposal.header.version.block` must be [`VERSION_BLOCK`](./consensus.md#constants).
+1. `proposal.header.version.app` must be [`VERSION_APP`](./consensus.md#constants).
+1. `proposal.header.height` should be previous known height + 1.
+1. `proposal.header.chain_id` must be [`CHAIN_ID`](./consensus.md#constants).
+1. `proposal.header.time` is processed identically to Tendermint.
+1. `proposal.header.last_header_hash` must be previous block's header hash.
+1. `proposal.header.last_commit_hash` must be the previous block's commit hash.
+1. `proposal.header.consensus_hash` must be the hash of [consensus parameters](./data_structures.md#header).
+1. `proposal.header.state_commitment` must be the state root after applying the previous block's transactions.
+1. `proposal.header.available_data_original_shares_used` must be at most [`AVAILABLE_DATA_ORIGINAL_SQUARE_MAX ** 2`](./consensus.md#constants).
+1. `proposal.header.available_data_root` must be the [root](./data_structures.md#availabledataheader) of `proposal.da_header`.
+1. `proposal.header.proposer_address` must be the [correct leader](./consensus.md#leader-selection).
+1. `proposal.da_header` must be well-formed.
+1. The number of elements in `proposal.da_header.row_roots` and `proposal.da_header.row_roots` must be equal.
+1. The number of elements in `proposal.da_header.row_roots` must be the same as computed [here](./data_structures.md#header).
+1. `proposal.proposer_signature` must be a valid [digital signature](./data_structures.md#public-key-cryptography) over the header hash of `proposal.header` that recovers to `proposal.header.proposer_address`.
+1. For [full nodes](./node_types.md#node-type-definitions), `proposal.da_header` must be the result of computing the roots of the shares (received separately).
+1. For [light nodes](./node_types.md#node-type-definitions), `proposal.da_header` should be sampled from for availability.
```

Comment:
This is dope!

In general, I'm in favor of a) stating there is a validity-predicate isValidProposal(...) and b) a pseudo-code listing that clarifies input of that predicate and lists validation steps in pseude-code instead.

I can provide that as a follow-up to this PR.

Reply:
I agree all those are good things. Would prefer deferring to a future PR though as essentially a proto-ADR (or an actual ADR?).

Reply:
Yeah, we can do it in followup PRs. Just noting that pseudo-code would be better independent of existing implementation infrastructure etc.